src/Pathy/Path.purs

```diff
@@ -0,0 +1,291 @@
+-- |   directory of the path it receives (eliminates `parentOf`).
+-- | - The third argument is a function representing a file or directory within
+-- |   the directory of the path it receives (eliminates `extendPath`).
+foldPath
+  :: forall a b r
+   . r
+  -> (Path a Dir -> r)
+  -> (Path a Dir -> Name b -> r)
+  -> Path a b
+  -> r
+foldPath r f g = case _ of
+  Init -> r
+  ParentOf d -> f d
+  In d n -> g d n
+
+-- | Peels off the last directory and the terminal file or directory name
+-- | from the path. Returns `Nothing` if the path is `rootDir` / `currentDir` or
+-- | some `parentOf p`.
+peel :: forall a b. Path a b -> Maybe (Tuple (Path a Dir) (Name b))
+peel = foldPath Nothing (const Nothing) (\p n -> Just (Tuple p n))
+
+-- | Peels off the last director and terminal file from a path. Unlike the
+-- | general `peel` function this is guaranteed to return a result, as `File`
+-- | paths are known to have a name.
+peelFile :: forall a. Path a File -> Tuple (Path a Dir) (Name File)
+peelFile = case _ of
+  Init -> unsafeCrashWith "`Init` in Pathy.peelFile (this should be impossible)"
+  ParentOf _ -> unsafeCrashWith "`ParentOf` in Pathy.peelFile (this should be impossible)"
+  In p n -> Tuple p n
+
+-- | Retrieves the name of the terminal segment in a path. Returns `Nothing` if
+-- | the path is `rootDir` / `currentDir` or some `parentOf p`.
+name :: forall a b. IsRelOrAbs a => IsDirOrFile b => Path a b -> Maybe (Name b)
+name = foldPath Nothing (const Nothing) (const Just)
+
+-- | Retrieves the name of a file path. Unlike the general `name` function,
+-- | this is guaranteed to return a result, as `File` paths are known to have a
+-- | name.
+fileName :: forall a. Path a File -> Name File
+fileName = case _ of
+  Init -> unsafeCrashWith "`Init` in Pathy.fileName (this should be impossible)"
+  ParentOf _ -> unsafeCrashWith "`ParentOf` in Pathy.fileName (this should be impossible)"
+  In _ n -> n
+
+-- | Attempts to rename the terminal segment of a path. If the path is
+-- | `rootDir` / `currentDir` or some `parentOf p` this will have no effect.
+rename :: forall a b. (Name b -> Name b) -> Path a b -> Path a b
+rename f = un Identity <<< renameTraverse (pure <<< f)
+
+-- | Attempts to rename the terminal segment of a path using a function that
+-- | returns the result in some `Applicative`. If the path is `rootDir` /
+-- | `currentDir` or some `parentOf p` this will have no effect.
+renameTraverse
+  :: forall f a b
+   . Applicative f
+  => (Name b -> f (Name b))
+  -> Path a b
+  -> f (Path a b)
+renameTraverse f = case _ of
+  In p n -> In p <$> f n
+  p -> pure p
+
+-- | Sets the extension of a name.
+-- |
+-- | ```purescript
+-- | file "image" <.> "png"
+-- | ```
+setExtension :: forall a b. Path a b -> NonEmptyString -> Path a b
+setExtension p ext = rename (alterExtension (const (Just ext))) p
+
+infixl 6 setExtension as <.>
+
+-- | Makes a path relative to a reference path.
```

Comment:
Order of arguments for this function is not that obvious would be nice to add  couple examples and maybe more explanation :d

Reply:
or if we add this property here that would also help
```
let cp = canonicalize p :: Path a b
let cpRoot = canonicalize pRoot :: Path  Abs Dir
cp == canonicalize (cpRoot </> cp `relativeTo` cpRoot)
```

Reply:
Examples 👍 

I don't know how to explain it better :smile:

Reply:
Yeah I think having the property explained is useful too. Can drop all the `canonicalize`s from there, that was more just to clear up any weirdness from the test output.

Reply:
✅  added the property